File: packages/python-sdk/src/urisocial/resources/images.py
"""Images resource"""

from typing import TYPE_CHECKING, Optional
from ..types import ImageGenerationRequest, ImageGenerationResult

if TYPE_CHECKING:
    from ..http_client import HTTPClient


class ImagesResource:
    """Resource for AI image generation and editing"""

    def __init__(self, http: "HTTPClient"):
        self._http = http

    def generate(self, **kwargs: ImageGenerationRequest) -> ImageGenerationResult:
        """
        Generate an image using AI

        Args:
            prompt: Image generation prompt
            reference_image: Optional product/reference image for preservation
            style: Image style (immersive, standard, minimalist)
            aspect_ratio: Image aspect ratio (1:1, 4:5, 16:9)

        Returns:
            Generated image URL and revised prompt

        Example:
            >>> image = client.images.generate(
            ...     prompt='Luxury perfume bottle on marble surface',
            ...     reference_image='https://example.com/product.jpg',
            ...     style='immersive',
            ...     aspect_ratio='1:1'
            ... )
        """
        return self._http.post("/api/v1/images/generate", json=kwargs)

    def edit(
        self,
        image_url: str,
        prompt: str,
        mask_url: Optional[str] = None,
    ) -> ImageGenerationResult:
        """
        Edit an existing image

        Args:
            image_url: URL of image to edit
            prompt: Edit instruction prompt
            mask_url: Optional mask image URL for targeted edits

        Returns:
            Edited image URL
        """
        return self._http.post(
            "/api/v1/images/edit",
            json={
                "image_url": image_url,
                "prompt": prompt,
                "mask_url": mask_url,
            },
        )

    def remove_background(self, image_url: str) -> dict:
        """
        Remove background from product image

        Args:
            image_url: URL of product image

        Returns:
            Dictionary with cutout_url

        Example:
            >>> result = client.images.remove_background(
            ...     'https://example.com/product.jpg'
            ... )
            >>> print(result['cutout_url'])
        """
        return self._http.post(
            "/api/v1/images/remove-background",
            json={"image_url": image_url},
        )

    def analyze_product(self, image_url: str) -> dict:
        """
        Analyze product in image (forensic analysis)

        Args:
            image_url: URL of product image

        Returns:
            Product analysis with type, shape, colors, materials, labels

        Example:
            >>> analysis = client.images.analyze_product(
            ...     'https://example.com/perfume.jpg'
            ... )
            >>> print(analysis['product_type'])
            >>> print(analysis['colors'])
        """
        return self._http.post(
            "/api/v1/images/analyze-product",
            json={"image_url": image_url},
        )

File: packages/typescript-sdk/src/resources/images.ts
import { HTTPClient } from '../client/http';
import { ImageGenerationRequest, ImageGenerationResult } from '../types';

export class ImagesResource {
  constructor(private http: HTTPClient) {}

  /**
   * Generate an image using AI
   */
  async generate(request: ImageGenerationRequest): Promise<ImageGenerationResult> {
    return this.http.post<ImageGenerationResult>('/api/v1/images/generate', request);
  }

  /**
   * @deprecated Not currently supported by the backend — there is no generic,
   * URL-based image edit endpoint. For editing images already attached to a
   * draft, regenerate the draft's image via the drafts/content workflow
   * instead. This method throws immediately rather than making a network
   * call that would 404.
   */
  async edit(_request: {
    imageUrl: string;
    prompt: string;
    maskUrl?: string;
  }): Promise<ImageGenerationResult> {
    throw new Error(
      'images.edit() is not supported by the URI Social API — there is no generic ' +
        'image edit endpoint. Regenerate the draft image instead.'
    );
  }

  /**
   * Remove background from product image
   */
  async removeBackground(imageUrl: string): Promise<{ cutout_url: string }> {
    return this.http.post<{ cutout_url: string }>('/api/v1/images/remove-background', {
      image_url: imageUrl,
    });
  }

  /**
   * Analyze product in image (forensic analysis)
   */
  async analyzeProduct(imageUrl: string): Promise<{
    product_type: string;
    shape: string;
    colors: string[];
    materials: string[];
    labels: string[];
  }> {
    return this.http.post('/api/v1/images/analyze-product', {
      image_url: imageUrl,
    });
  }
}
